### ISSUE TITLE
a trouble when testing cyclegan

### ISSUE THREAD

**Issue by bearxiong333** [junyanz/pytorch-CycleGAN-and-pix2pix#1122]
Hi, appreciating your open source code, it's really a masterpiece.

I've met some trouble when I run the test.py using cyclegan, like this:
![image](https://user-images.githubusercontent.com/69395938/89719307-f78fb900-d994-11ea-990b-f68e9b68708f.png)

my input data are images with [256,256,3], I keep some flags the same as training, such as netG, norm, dropout, like belows:
!python train.py --dataroot /content/data/single/ --netG unet_256 --batch_size 64 --model cycle_gan1 --gpu_ids 0 --serial_batches --preprocess none --no_flip --name XQ12_singlecycle_noidentity --n_epochs 50 --n_epochs_decay 50 --print_freq 3200 --save_epoch_freq 10 --save_latest_freq 12096 --update_html_freq 40000 --lambda_identity 0.0 --lambda_supervised 0.0 --lambda_per 0.0 --lambda_A 10 --lambda_B 10 --beta_A 10 --beta_B 10 --perceptual_layers 8 --continue_train --epoch_count 61

!python test.py --dataroot /content/data/single/ --batch_size 1 --netG unet_256 --model testcyclegan1 --gpu_ids 0 --serial_batches --preprocess none --no_flip --name XQ12_singlecycle_noidentity --input_nc 3 --no_dropout --eval --dataset_mode singlenew1 --results_dir /content/data/results1_singlecycle/ --num_test 14382 --model_suffix _A

please don't mind some strange flags, I just add these flags for some new loss functions based on your wonderful code.
I printed out the tensor size before this problem appeared, the size is [1,512,2,2], which I think is OK. Actually, I met this problem once before, that happened in the training stage, the reason is I used BN in my network and I set my batch_size as 1. I searched this problem on the internet, someone told me if using BN, the batch_size must >1. But in the cyclegan, we use IN right? I tried to change batch_size in the test.py to 2 which was 1 as default. Then, I would not meet this problem, but there is a new one, the results  would skip a half of images. I can only get image0,2,4,6... without image1,3,5,7... . Could you please help me find out the solution? Thanks.

**Comment by bearxiong333** [comment]
I guess I know why I would miss a half of images when I set batch_size =2, as picture shows:
![image](https://user-images.githubusercontent.com/69395938/89720550-33c91680-d9a1-11ea-9fb5-891eaa8eaa1f.png)
when I set batch_size = 2, that means the image_path would have two elements, but the code just use the first element by image_path[0]. I would try to make some change to see if I could save all images.

**Comment by junyanz** [comment]
Could you add `--norm instance` to both training and test scripts? 
You mentioned," the results would skip half of the images"?  Are you referring to the test script? Currently, the test script only supports batch_size=1.

**Comment by bearxiong333** [comment]
> Could you add `--norm instance` to both training and test scripts?
> You mentioned," the results would skip half of the images"? Are you referring to the test script? Currently, the test script only supports batch_size=1.

Thank you for your attention. Yes, currently, we could only use batch_size=1 in the test script. I checked the code where referring the norm layer several times, I am sure the norm is instance normalization not batch normalization. Moreover, the option list which would automatically prints out when running the code also shows  'norm: instance' no matter as train stage or test stage. Now, I try to annotate the two related lines of codes which cause this problem in the .../torch/nn/functional.py. It seems work, but I am not sure whether it would affect the quality of results.

**Comment by junyanz** [comment]
Even for `--norm instance`, given a BxCxHxW tensor, the layer should assume that HxW > 1. It may not work for 1x512x1x1.

**Comment by bearxiong333** [comment]
> Even for `--norm instance`, given a BxCxHxW tensor, the layer should assume that HxW > 1. It may not work for 1x512x1x1.

your answer reminders me. Perhaps, I know why I met this problem but you didn't
Actually, I referred another reference (https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=8825802)and made a little change to original cyclegan network. 
![image](https://user-images.githubusercontent.com/69395938/89724792-83293a00-d9d5-11ea-9200-bcf1badb2e11.png)
As you see (red rectangle), the reference adds a normalization layer (BN) in the encode of the innermost. The problem is this normalization layer, because the tensor must be Bx512x1x1 after the last downconvolution, here HxW=1, so your original code doesn't have normalization layer in this place, which makes sense. However, in the train stage, I set batch_size>1 , the problem would not appear. So, can I assume if B>1, the normalization layer could accept HxW=1? If this assumption is right, it could explain why I would meet this problem at test stage, because when B=1 and HxW=1, I still put the tensor into a normalization layer which is different from you original code.

**Comment by bearxiong333** [comment]
I'am still confused about my assumption above. In the train stage, I also use instance normalization not BN. When I set batch_size >1, it can be trained well. But the point is IN has nothing to do with B and C dimensions, right? That is to say, no matter the B =1 or B>1, it would not affect IN, so why when I set batch_size>1, this problem would disappear? doesn't make sense.

**Comment by junyanz** [comment]
from the error message, you got a tensor with shape 1x512x1x1. This will break InstanceNorm as both H and W =1.

**Comment by bearxiong333** [comment]
> from the error message, you got a tensor with shape 1x512x1x1. This will break InstanceNorm as both H and W =1.

yes, thank you for your help~~
